autopilot/hostscore.go

```diff
@@ -0,0 +1,263 @@
+	// This ratio should be used by checking that the MinSectorAccessPrice is
+	// less than or equal to the MinDownloadBandwidthPrice multiplied by this
+	// constant
+	maxSectorAccessPriceVsBandwidth = uint64(400e3)
+)
+
+func hostScore(cfg Config, h Host) float64 {
+	// TODO: priceAdjustmentScore
+	// TODO: storageRemainingScore
+	return ageScore(h) *
+		collateralScore(cfg, h) *
+		interactionScore(h) *
+		settingsScore(cfg, h) *
+		uptimeScore(h) *
+		versionScore(h)
+}
+
+func ageScore(h Host) float64 {
+	if len(h.Announcements) == 0 {
+		return math.SmallestNonzeroFloat64
+	}
+
+	const day = 24 * time.Hour
+	weights := []struct {
+		age    time.Duration
+		factor float64
+	}{
+		{128 * day, 1.5},
+		{64 * day, 2},
+		{32 * day, 2},
+		{16 * day, 2},
+		{8 * day, 3},
+		{4 * day, 3},
+		{2 * day, 3},
+		{1 * day, 3},
+	}
+
+	age := time.Since(h.Announcements[0].Timestamp)
+	weight := 1.0
+	for _, w := range weights {
+		if age >= w.age {
+			break
+		}
+		weight /= w.factor
+	}
+
+	return weight
+}
+
+func collateralScore(cfg Config, h Host) float64 {
+	settings, _, ok := h.LastKnownSettings()
+	if !ok {
+		return math.SmallestNonzeroFloat64
+	}
+
+	// NOTE: This math is copied directly from the old siad hostdb. It would
+	// probably benefit from a thorough review.
+
+	// NOTE: We are not multiplying the `Storage` with redundancy, which is
+	// different from the siad implementation
+
+	// convenience variables
+	allowance := cfg.Contracts.Allowance
+	numhosts := cfg.Contracts.Hosts
+	duration := cfg.Contracts.Period
+	storage := cfg.Contracts.Storage
+
+	// calculate the collateral
+	contractCollateral := settings.Collateral.Mul64(storage).Mul64(duration)
+	contractCollateralMax := settings.MaxCollateral.Div64(2) // 2x buffer - renter may end up storing extra data
+	if contractCollateral.Cmp(contractCollateralMax) > 0 {
+		contractCollateral = contractCollateralMax
+	}
+	collateral, _ := new(big.Rat).SetInt(contractCollateral.Big()).Float64()
+	collateral = math.Max(1, collateral) // ensure collateral is at least 1
+
+	// calculate the cutoff
+	expectedFundsPerHost := allowance.Div64(numhosts)
+	cutoff, _ := new(big.Rat).SetInt(expectedFundsPerHost.Div64(5).Big()).Float64()
+	cutoff = math.Max(1, cutoff)          // ensure cutoff is at least 1
+	cutoff = math.Min(cutoff, collateral) // ensure cutoff is not greater than collateral
+
+	// calculate the weight
+	ratio := collateral / cutoff
+	smallWeight := math.Pow(cutoff, 4)
+	largeWeight := math.Pow(ratio, 0.5)
+	weight := smallWeight * largeWeight
+	return weight
+}
+
+func interactionScore(h Host) float64 {
+	success, fail := 30.0, 1.0
+	for _, hi := range h.Interactions {
+		if hi.Success {
+			success++
+		} else {
+			fail++
+		}
+	}
+
+	weight := math.Pow(success/(success+fail), 10)
+	return weight
+}
+
+func settingsScore(cfg Config, h Host) float64 {
+	settings, _, found := h.LastKnownSettings()
+	if !found {
+		return math.SmallestNonzeroFloat64
```

Comment:
Using `math.SmallestNonzeroFloat64` has always bugged me. I'd prefer to return 0 and filter out those hosts before selecting (to avoid divide-by-zero in `randSelectByWeight`)